### ISSUE TITLE
MSVC 2022: C++ library build broken with UNICODE

### ISSUE THREAD

**Issue by gpascualg** [dfeneyrou/palanteer#29]
When trying to build the c++ library while having UNICODE declared, the following error occurs:

```
palanteer-src\c++\palanteer.h(4346,53): error C2440: '=': cannot convert from 'const wchar_t [17]' to 'LPWSTR'
```

Basically, in that line, `LogFile.LoggerName          = KERNEL_LOGGER_NAME;` the structure field expects a non-const value, while `KERNEL_LOGGER_NAME` is const. I have worked around it by casting it to `(LPWSTR)`. I don't know the API exactly, so it feels like a hack to cast it to non-constness.

```diff
- LogFile.LoggerName          = KERNEL_LOGGER_NAME;
+ #if defined(_UNICODE) || defined(UNICODE)
+ LogFile.LoggerName          = (LPWSTR)KERNEL_LOGGER_NAME;
+ #else
+ LogFile.LoggerName          = KERNEL_LOGGER_NAME;
+ #endif
```

**Comment by dfeneyrou** [comment]
I was unable to reproduce this compilation issue with MSVC 2019 nor MSVC 2022 (installed it on purpose, only the "Build Tools"), when trying to build the test program.

In ./c++/testprogram/CMakeList.txt, the `-DUNICODE` is added by default, so I thought that it was linked to MSVC 2022. 
With or without the addition of `-DUNICODE` , the type of the system constant `KERNEL_LOGGER_NAME` changes (`LPSTR` or `LPWSTR`) but is always consistent with the `LoggerName`  type and I never got the `const` issue that you faced.

In short, in my case, both the actual code and your patch are working, also both for MSVC 2019 or 2022.
I used the command line `nmake testprogram` for all tests.

Can you give more details on your build setup and/or reproduce the issue with the provided test program (so with `nmake testprogram`)?

**Comment by gpascualg** [comment]
I'm currently away from home, but this Monday when I get back I'll build it and send you the outputs.
Thank you!

**Comment by BrodyHiggerson** [comment]
I've had the same issue in VS2019 v16.8.3 alongside a slew of warnings (will dig those out and open a separate issue listing them / a PR fixing them).

I'm using these flags:

/W4 /WX /std:c++latest /permissive- /D WIN32_LEAN_AND_MEAN /D NOMINMAX

Wonder if it's the permissive- (i.e. strict mode) flag.

**Comment by gpascualg** [comment]
Running `nmake testprogram` as is does, indeed, work. I have however isolated two instances that cause the error I was getting.

1) As @BrodyHiggerson says, adding `/permissive-` will trigger the error. I suspect the error is always there, just that without permissive the cast is being done implicitely.

2) Adding `/std:c++20`, which is my case. It seems enabling c++20 automatically enables strict conformance (permissive-).
https://devblogs.microsoft.com/cppblog/msvc-cpp20-and-the-std-cpp20-switch/

It definitely seems to be caused by strict mode.

**Comment by dfeneyrou** [comment]
Thanks, I was able to reproduce it with /W4 /permissive- indeed!
